DOCS 918: Improve service api docs (#1508)
viamrobotics/docs: docs/services/slam/_index.md

| Model | Description |
| ----- | ----------- |
| [`viam:slam:cartographer`](cartographer/) | [The Cartographer Project](https://github.com/cartographer-project) performs dense SLAM using LIDAR data. |

## API

The SLAM Service supports the following methods:

Method Name | Description
----------- | -----------

> [`GetPosition`](#getposition) | Get the current position of the specified source component in the point cloud SLAM map.

Reviewer: [note] Once we support multiple sensor inputs, we'll want to come back and update this description and similar descriptions below.

cc @PStrutz , @jeremyrhyde. Also, for Patricia and Jeremy to think about - with IMU support, does it still make sense for GetPosition to return a single componentReference?

**Returns:**

- [(spatialmath.Pose)](https://pkg.go.dev/go.viam.com/rdk@v0.3.0/spatialmath#Pose): A `Pose` representing the current position of the specified component.
- [(string)](https://pkg.go.dev/builtin#string): The `"name"` of the component the SLAM service is configured to source point cloud data from.
For example, a [camera](/components/camera/) named `"cam"`.
- [(error)](https://pkg.go.dev/builtin#error): An error, if one occurred.

For more information, see the [Go SDK Docs](https://pkg.go.dev/go.viam.com/rdk/services/slam).

```go {class="line-numbers linkable-line-numbers"}
slam_svc, err := slam.FromRobot(robot, "my_slam_service")

// Get the current position of the specified source component in the SLAM map as a Pose.
pos, name, err := slam_svc.GetPosition(context.Background())
```

{{% /tab %}}
{{% tab name="Python" %}}

**Parameters:**

- `timeout` [(Optional\[float\])](https://docs.python.org/library/typing.html#typing.Optional): An option to set how long to wait (in seconds) before calling a time-out and closing the underlying RPC call.

**Returns:**

> - [(Pose)](https://python.viam.dev/autoapi/viam/components/arm/index.html#viam.components.arm.Pose): A `Pose` representing the current position of the component the SLAM service is configured to source point cloud data from.

Reviewer: @sguequierre , @jeremyrhyde is it right to link to components.arm.Pose, or is there a slam-specific link that would make more sense?

> - [(Pose)](https://python.viam.dev/autoapi/viam/components/arm/index.html#viam.components.arm.Pose): A `Pose` representing the current position of the component the SLAM service is configured to source point cloud data from.

Author: ```suggestion
- [(Pose)](https://python.viam.dev/autoapi/viam/services/slam/index.html#viam.services.slam.Pose): A `Pose` representing the current position of the component the SLAM service is configured to source point cloud data from.
```

For example, a [camera](/components/camera/) named `"cam"`.

For more information, see the [Python SDK Docs](https://python.viam.dev/autoapi/viam/services/slam/client/index.html#viam.services.slam.client.SLAMClient.get_position).

```python {class="line-numbers linkable-line-numbers"}
slam_svc = SLAMClient.from_robot(robot=robot, name="my_slam_service")

# Get the current position of the specified source component in the SLAM map as a Pose.
pose = await slam.get_position()
```

{{% /tab %}}
{{< /tabs >}}

### GetPointCloudMap

Get the point cloud map.

{{< tabs >}}
{{% tab name="Go" %}}

**Parameters:**

- `ctx` [(Context)](https://pkg.go.dev/context): A Context carries a deadline, a cancellation signal, and other values across API boundaries.

**Returns:**

> - (func() [[]byte](https://pkg.go.dev/builtin#byte), [error](https://pkg.go.dev/builtin#error)): The complete point cloud map in standard PCD format.

Reviewer: @sguequierre is it right that `[error](https://pkg.go.dev/builtin#error))` should be on this line? I wasn't sure since the line description ("The complete point cloud map in standard PCD format") just refers to the `[]byte` part, and there's a line below about the error part

Author: I agree this is a bit confusing, I meant it to match the way the function return is typed here-- https://pkg.go.dev/go.viam.com/rdk/services/slam#Service -- do you think that's ok?

Reviewer: Ohh I see, you're right! It's returning a function that takes no parameters and returns the byte array or an error. Resolving this and the other thread.

- `timeout` [(Optional\[float\])](https://docs.python.org/library/typing.html#typing.Optional): An option to set how long to wait (in seconds) before calling a time-out and closing the underlying RPC call.

**Returns:**

- [(List[GetPointCloudMapResponse])](https://python.viam.dev/autoapi/viam/gen/service/slam/v1/slam_pb2/index.html#viam.gen.service.slam.v1.slam_pb2.GetPointCloudMapResponse): The complete point cloud map in standard PCD format.

For more information, see the [Python SDK Docs](https://python.viam.dev/autoapi/viam/services/slam/client/index.html#viam.services.slam.client.SLAMClient.get_point_cloud_map).

```python {class="line-numbers linkable-line-numbers"}
slam_svc = SLAMClient.from_robot(robot=robot, name="my_slam_service")

# Get the point cloud map in standard PCD format.
pcd_map = await slam.get_point_cloud_map()
```

{{% /tab %}}
{{< /tabs >}}

### GetInternalState

Get the internal state of the SLAM algorithm required to continue mapping/localization.

{{< tabs >}}
{{% tab name="Go" %}}

**Parameters:**

- `ctx` [(Context)](https://pkg.go.dev/context): A Context carries a deadline, a cancellation signal, and other values across API boundaries.

> - `extra` [(map\[string\]interface{})](https://go.dev/blog/maps): Extra options to pass to the underlying RPC call.

Reviewer: @sguequierre should "extra" be here? It's not in the [slam.proto](https://github.com/viamrobotics/api/blob/03915ccb108a2284ea0083edc8c8edffc4f825b0/proto/viam/service/slam/v1/slam.proto#L97-L100) and in the example below GetInternalState just takes a `ctx` argument

Author: no you're absolutely right, I did not catch this when trying to QA yesterday-- thank you!

Author: will fix

> - `extra` [(map\[string\]interface{})](https://go.dev/blog/maps): Extra options to pass to the underlying RPC call.

Author: ```suggestion
```

**Returns:**

> - (func() [[]byte](https://pkg.go.dev/builtin#byte), [error](https://pkg.go.dev/builtin#error)): Chunks of the internal state of the SLAM algorithm.

Reviewer: @sguequierre Same question about error here

Author: I think here the way it renders with the links makes it look extra off but I meant it to be the way the return is typed here https://pkg.go.dev/go.viam.com/rdk/services/slam#Service, am I misunderstanding?

# Get the internal state of the SLAM algorithm required to continue mapping/localization.
internal_state = await slam.get_internal_state()
```

{{% /tab %}}
{{< /tabs >}}

### GetLatestMapInfo

{{% alert title="Important" color="tip" %}}

This method is not yet available in the Viam Python SDK.

{{% /alert %}}

Get the timestamp of the last update to the point cloud SLAM map.

{{< tabs >}}
{{% tab name="Go" %}}

**Parameters:**

- `ctx` [(Context)](https://pkg.go.dev/context): A Context carries a deadline, a cancellation signal, and other values across API boundaries.

**Returns:**

- [(time.Time)](https://pkg.go.dev/time#Time): The timestamp of the last update to the point cloud SLAM map.
- [(error)](https://pkg.go.dev/builtin#error): An error, if one occurred.

For more information, see the [Go SDK Docs](https://pkg.go.dev/go.viam.com/rdk/services/slam#Service).

```go {class="line-numbers linkable-line-numbers"}
slam_svc, err := slam.FromRobot(robot, "my_slam_service")

// Get the timestamp of the last update to the point cloud SLAM map.
timestamp, err := slam_svc.GetLatestMapInfo(context.Background())
```

{{% /tab %}}

> {{< /tabs >}}

Author: ```suggestion
{{< /tabs >}}

```
